AlgebraInLean/Chapter02/Sheet01.lean

```diff
@@ -3,252 +3,90 @@ import AlgebraInLean.Chapter02.Sheet00
-    _ = μ (μ (ι a) a) c := by rw [op_assoc]
-    _ = μ 𝕖 c := by rw [inv_op a]
-    _ = c := by rw [id_op]
-  -- END SAMPLE SOLUTION
-/-
-
-This is a familiar proof from Chapter 1, but this time we're using the very nifty tactic `calc`.
-
-`calc`, like `have`, is an example of what is called "forward reasoning" in Lean. Usually with
-tactic-style proofs, we are trying to go backwards, transforming our goal to one of our assumed
-hypotheses. With forward reasoning, we are trying to transform our assumptions into the goal.
-
-`calc` might be more easier to work with than you think, since it closely reflects some
-pen-and-paper proofs via a chain of equalities and algebraic manipulations. Hover over `calc` to see
-the syntax. Don't let it scare you!
-
--/
-
-/- For all g ∈ G, (g⁻¹)⁻¹ = g -/
-theorem inv_inv_eq_self : ∀ g : G, ι (ι g) = g := by
-  -- SAMPLE SOLUTION
-  intro g
-  have hq : ∀ (a : G), μ (ι a) a = μ a (ι a)
-  · intro a
-    rw [inv_op a]
-    rw [op_inv a]
-  specialize hq (ι g)
-  rw [inv_op (ι g)] at hq
-  symm at hq
-  rw [← inv_op g] at hq
-  rw [mul_left_eq (ι g) (ι (ι g)) g hq]
-  -- END SAMPLE SOLUTION
-
-/- For all a, b ∈ G, a⁻¹ = b⁻¹ → a = b -/
-example : ∀ a b : G, ι a = ι b → a = b := by
-  -- SAMPLE SOLUTION
-  intro a b
-  intro hinv
-  have hinj : ∀ (g : G), ι (ι g) = g
-  · apply inv_inv_eq_self
-  rw [← hinj a, ← hinj b]
-  rw [hinv]
-  -- END SAMPLE SOLUTION
-
-/- The inverse function is injective -/
-theorem inv_inj : Injective (ι: G → G) := by
-  -- SAMPLE SOLUTION
-  unfold Injective
-  have hinv : ∀ (x : G), ι (ι x) = x
-  · intro x
-    rw [inv_inv_eq_self x]
-  intro a b hab
-  rw [← hinv a, ← hinv b, hab]
-  -- END SAMPLE SOLUTION
-/-
-
-`unfold` does what it sounds like: unfolding a symbol to its underlying definition. It isn't best
-practice; it's usually better to write a definition to use `rw` with. However, for a one-off
-use-case, `unfold` suffices.
-
-You saw the following examples in Sheet 0, but in a much different way. It may be useful to review
-different approaches for the following proofs:
+Group homomorphisms are a particular case of homomorphisms in algebra, which are maps between
+algebraic structures (like monoids or semigroups) that preserve the algebraic structure like this.
+These are a kind of an even more general concept of a "morphism" from category theory:
+<https://en.wikipedia.org/wiki/Category_theory>.
 
+Mathematicians often omit the structure of the morphism (or homomorphism) when it can be deduced
+from context, so these will often be called just "homomorphism" rather than "group homomorphism".
 -/
 
-variable {α β γ : Type*}
+variable {G H K : Type*} [Group G] [Group H] [Group K]
 
-/- An injective and surjective function is bijective -/
-example (f : α → β) (h1 : Injective f) (h2 : Surjective f)
-: (Bijective f) := by
-  rw [Bijective]
-  constructor <;> assumption
+def Group.Homomorphism (φ : G → H) : Prop := ∀ a b : G, μ (φ a) (φ b) = φ (μ a b)
 
 /-
-
```

Comment:
See above comment ^

Reply:
We already discussed `all_goals` before, so I there isn't really a need to explain `<;>`, especially in such depth. Maybe I should mention that it can serve as a shorter alternative somewhere.